src/lib/storage/index/partial_hash/partial_hash_index.hpp

```diff
@@ -0,0 +1,154 @@
+ * additional chunks or removing already indexed chunks.
+ */
+class PartialHashIndex {
+  friend PartialHashIndexTest;
+
+ public:
+  using Iterator = BasePartialHashIndexImpl::Iterator;
+  using IteratorRange = BasePartialHashIndexImpl::IteratorRange;
+  using IteratorRangePair = BasePartialHashIndexImpl::IteratorRangePair;
+
+  PartialHashIndex() = delete;
+  PartialHashIndex(const std::vector<std::pair<ChunkID, std::shared_ptr<Chunk>>>& chunks_to_index, const ColumnID);
+
+  /**
+   * The following four methods are used to access any table index. Each of them accepts a generic function object that
+   * expects a begin and end iterator to the underlying data structure as parameter. When accessing the index with one
+   * of these methods, one or two pairs of iterators are passed to the function object. For more information, please
+   * refer to the method description.
+   * 
+   * Unfortunately, it is not possible to locate the implementations of these methods inside the source file because of
+   * the separate compilation model. In this case, it is not possible to use explicit template instantiation since
+   * otherwise all used function objects would have to be instantiated in abstract_table_index.cpp.
+   */
+
+  /**
+   * Acquires iterators to the first and last indexed non-NULL elements and passes them to the passed functor.
+   * 
+   * @param functor is a generic function object accepting two iterators as arguments
+  */
+  template <typename Functor>
+  void access_values_with_iterators(const Functor& functor) const {
+    const auto lock = std::shared_lock<std::shared_mutex>(_data_access_mutex);
+    functor(_cbegin(), _cend());
+  }
+
+  /**
+   * Acquires iterators to the first and last indexed NULL elements and passes them to the passed functor.
+   * 
+   * @param functor is a generic function object accepting two iterators as arguments
+  */
+  template <typename Functor>
+  void access_null_values_with_iterators(const Functor& functor) const {
+    const auto lock = std::shared_lock<std::shared_mutex>(_data_access_mutex);
+    functor(_null_cbegin(), _null_cend());
+  }
+
+  /**
+   * Searches for all positions of the entry within the table index and acquires a pair of Iterators containing the
+   * start and end iterator for the stored RowIDs of the element inside the table index. These are then passed to the
+   * functor.
+   * 
+   * @param functor is a generic function object accepting two iterators as arguments
+   * @param value is the entry searched for
+  */
+  template <typename Functor>
+  void range_equals_with_iterators(const Functor& functor, const AllTypeVariant& value) const {
+    const auto lock = std::shared_lock<std::shared_mutex>(_data_access_mutex);
+    const auto& [index_begin, index_end] = _range_equals(value);
+    functor(index_begin, index_end);
+  }
+
+  /**
+   * Searches for all positions that do not equal to the entry in the table index and acquires a pair of IteratorRanges
```

Comment:
"Searches for all positions that *are* not equal to the entry in the table index..."?